qutip/tests/test_enr_state_operator.py

```diff
@@ -197,3 +198,101 @@ def test_eigenstates_ENR():
     vals, vecs = H.eigenstates()
     for val, vec in zip(vals, vecs):
         assert val * vec == H @ vec
+
+
+# function to convert an ENR Qobj to a regular Qobj
+def enr_to_reg(q):
+    if q.type == "scalar":
+        return q
+    mat = q.data_as().tocoo()
+    dims = q.dims[0]
+    n_ex = q._dims[0].n_excitations
+
+    if q.isoper:
+        newdims = [list(dims)]*2
+        newshape = [np.prod(dims)]*2
+    elif q.isket:
+        newdims = [list(dims), [1]*len(dims)]
+        newshape = [np.prod(dims), 1]
+    else:
+        raise ValueError("Invalid Qobj type")
+
+    newindices = [i for i, state  # get indices for unrestricted space
+                  in enumerate(qutip.state_number_enumerate(dims))
+                  if sum(state) <= n_ex]
+    newrows = [newindices[i] for i in mat.row]
+    newcols = [newindices[i] for i in mat.col]
+    newmat = scipy.sparse.coo_matrix(
+        (mat.data, (newrows, newcols)), shape=newshape)
+
+    return qutip.Qobj(newmat, dims=newdims, dtype="csr")
+
+
+# copied parameters from test_ptrace.py
+@pytest.mark.parametrize('dims, sel',
+                         [
+                             ([5, 2, 3], [2, 1]),
+                             ([5, 2, 3], [0, 2]),
+                             ([5, 2, 3], [0, 1]),
+                             ([2]*6, [3, 2]),
+                             ([2]*6, [0, 2]),
+                             ([2]*6, [0, 1]),
+                             ([4, 4, 4], []),
+                             ([4, 4, 4], [0, 1, 2]),
+                         ])
+def test_enr_ptrace(dims, sel, n_excitations):
+    nstates_enr = _n_enr_states(dims, n_excitations)
+    # use qutip to make a random sparse Hermitian matrix w/ trace 1
+    rand_dm_mat = qutip.rand_dm(
+        nstates_enr, 0.05, distribution="herm", dtype="csr").data_as()
+
+    rho_in_enr = qutip.Qobj(rand_dm_mat,
+                            dims=[qutip.energy_restricted.EnrSpace(
+                                dims, n_excitations)]*2)
```

Comment:
Why create a Qobj, extract the matrix and rebuild a new one?
You can update the dims of a Qobj:
`rand_dm_mat.dims = [qutip.energy_restricted.EnrSpace(dims, n_excitations)]*2`.

Ideally `qutip.rand_dm(EnrSpace(...), ...)` should work, we should fix...